luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,160 @@
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+
+	if unitDef.canMove then
+		mobileUnits[unitDefID] = true
+	end
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+	gadgetHandler:RegisterAllowCommand(CMD.STOP)
+end
+
+function gadget:GameFrame(frame)
+	local tagUpdates = 0
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if not commandData.cmdTag then
+						if tagUpdates >= tagUpdateFrameMax then break end
+
+						if not cmdCache[nanoID] then
+							cmdCache[nanoID] = spGetUnitCommands(nanoID, -1)
+							tagUpdates = tagUpdates +1
+						end
+
+						for _, cmd in ipairs(cmdCache[nanoID]) do
+							trackingTable[nanoID][cmd.params[1]].cmdTag = cmd.tag
+						end
+					end
+
+					local distance = spGetUnitSeparation(nanoID, targetID, constructionTurretsDefs[nanoDefID].buildRange3D, false)
+
+					if distance < maxDistance then
+						if commandData.status == statuses.outOfRange then
+							commandData.status = statuses.inRange
+						end
+					else
+						-- Do not blindly remove orders on out-of-range mobile units, as the player may have pre-emptively issued an order before the unit was in range
+						-- Instead, only remove the order once the unit has come into range then left again
+						if commandData.status == statuses.inRange then
+							spGiveOrderToUnit(nanoID, CMD.REMOVE, commandData.cmdTag, 0)
+							targets[targetID] = nil
+						end
+					end
+				else
+					targets[targetID] = nil
+				end
+			end
+		end
+	end
+end
+
+function gadget:AllowCommand(unitID, unitDefID, teamID, cmdID,	cmdParams, cmdOptions, cmdTag, synced, fromLua)
+	local constructionTurretDef = constructionTurretsDefs[unitDefID]
+	if not constructionTurretDef then return true end
+
+	if cmdID == CMD_STOP then
+		trackingTable[unitID] = nil
+	end
+
+	-- Rather than trying to handle area commands, let them fallthrough, then catch the resulting list of single-target commands
+	if #cmdParams ~= 1 then return true end
+
+	local distance = math.huge
+	local targetId = cmdParams[1]
+
+	if targetId < gMaxUnits then
+		if not spValidUnitID(targetId) then return end
+		local defID = spGetUnitDefID(targetId)
+
+		if mobileUnits[defID] then
+			trackingTable[unitID] = trackingTable[unitID] or {}
+			trackingTable[unitID][targetId] = {status = statuses.outOfRange}
```

Comment:
Much better to have an array instead of table here, in lua that syntax sugar can be really taxing. You could use locals too so the index is not just a number too like (`trackingTable[unitID][targetId][STATUS]`).

Sucks a bit but its the lua way imo, they don't design for having beautiful structures and objects.

Reply:
I think he means to replace the inner hash map with an array. `{[0]:CmdTag, [1]:status}` so Lua those not hash the index. And for readability access it with static non mut locals like STATUS and CMD_TAG, or well the camelCase equivalent. Because hashes are slow.

Reply:
Write is using the static locals, but access would be via `trackingTable[unitID][targetId][0]` and `trackingTable[unitID][targetId][1]`, which are completely meaningless to someone reading the code.

If this is such a performance hog that we _have_ to do it, then we need a postprocessor to convert our code. I am not willing to work with source code where I have to divine magic numbers everywhere.

Reply:
```
local TRACK_CMDTAG = 1
local TRACK_STATUS = 2

trackingTable[unitID][targetId][TRACK_CMDTAG] and trackingTable[unitID][targetId][TRACK_STATUS]
```

that's all you need

I agree with the postprocessor comment, maybe one day that will be possible, anyways the above is not beautiful but avoids the magic numbers problems for this kind of situations.